File: frontend/composables/ffapi.ts
import { Configuration, FFApi, type InConfigID } from "~/backend";
import type { FFConfig } from "~/backend/models/FFConfig";
import type { InFFCreate } from "~/backend/models/InFFCreate";
import type { InFFUpdate } from "~/backend/models/InFFUpdate";
import type { InSetState } from "~/backend/models/InSetState";
import type { OutGetAllFFConfig } from "~/backend/models/OutGetAllFFConfig";
import type { OutGetFFStatus } from "~/backend/models/OutGetFFStatus";
import type { State } from "~/backend/models/State";
import type { VideoAdaptationSet } from "~/backend/models/VideoAdaptationSet";
import type { VideoRepresentation } from "~/backend/models/VideoRepresentation";

export const useFFApi = () => {

    const configs = useFFConfigs();
    const status = useFFStatus();
    const runtimeConfig = useRuntimeConfig()
    const ffApi = new FFApi(new Configuration({basePath: runtimeConfig.public.apiBase}))

    async function create(serverUID: string, title: string, output: string): Promise<string> {
        const request: InFFCreate = {
            serverUid: serverUID,
            name: title,
            output: output
        };

        const uid: string = await ffApi.create({inFFCreate: request});

        await refresh_configs();
        await refresh_status();

        return uid;
    }

    async function remove(server_id: string, config_id: string): Promise<void> {
        const request: InConfigID = {
            configUid: config_id,
            serverUid: server_id,
        };

        await ffApi.remove({inConfigID: request});
        await refresh_configs();
        await refresh_status();
    }


    async function setState(server_uid: string, configUid: string, state: State) {
        const request: InSetState = {
            id: {
                serverUid: server_uid,
                configUid: configUid
            },
            state: state,
        };

        await ffApi.setConfigState({inSetState: request});
        await refresh_configs();
        await refresh_status();
    }

    async function update(config: FFConfig) {

        const request: InFFUpdate = {
            config: config,
        };

        await ffApi.update({inFFUpdate: request});
        await refresh_configs();
        await refresh_status();
    }

    async function refresh_configs() {
            const response: OutGetAllFFConfig = await ffApi.getAllConfigs();
            configs.value = response.configs;
    }

    async function getFFCmd(server_uid: string, configUid: string): Promise<string> {
        const request: InConfigID = {
            serverUid: server_uid,
            configUid: configUid
        };
        const response: string = await ffApi.getFfCommand({inConfigID: request});
        return response;
    }

    async function createDefaultVideoAdaptationSet() {
        const adaptationSet: VideoAdaptationSet = await ffApi.createDefaultVideoAdaptationSet();
        return adaptationSet;
    }

    async function createDefaultVideoRepresentation() {
        const representation: VideoRepresentation = await ffApi.createDefaultVideoRepresentation();
        return representation;
    }

    function getConfig(uid: string): FFConfig | undefined {
        return configs.value.find((config) => config.uid === uid);
    }

    async function refresh_status() {
    const response: OutGetFFStatus = await ffApi.getAllStatus();
        status.value = response.status;
    }

    return { create, remove, setState, update, refresh_configs, getFFCmd, createDefaultVideoAdaptationSet, createDefaultVideoRepresentation, getConfig, refresh_status };
}

File: backend/src/ff/ffapi.rs
use crate::core;
use crate::model;
use crate::model::ff::videoadaptationset::VideoAdaptationSet;
use crate::model::ff::videoadaptationset::VideoRepresentation;
use actix_web::dev::HttpServiceFactory;
use actix_web::get;
use actix_web::{post, web, Responder};
use utoipa_actix_web::OpenApiFactory;

#[utoipa::path(
    tags = ["FF"],
    responses(
        (status = 200, description = "OK", body = bool),
    )
)]
#[post("/update")]
pub async fn update(
    data: web::Json<model::ff::InFFUpdate>,
    core: web::Data<core::Core>,
) -> actix_web::Result<impl Responder> {
    core.ffdb
        .put_ff_config(data.config.clone())
        .await
        .map_err(|e| {
            actix_web::error::ErrorInternalServerError(format!("Fail to update config: {}", e))
        })?;
    Ok(web::Json(true))
}

#[utoipa::path(
    tags = ["FF"],
    responses(
        (status = 200, description = "OK", body = bool),
    )
)]
#[post("/set_config_state")]
pub async fn set_config_state(
    data: web::Json<model::InSetState>,
    core: web::Data<core::Core>,
) -> actix_web::Result<impl Responder> {
    core.ffdb
        .set_ff_config_state(&data.id.server_uid, &data.id.config_uid, data.state)
        .await
        .ok();
    Ok(web::Json(true))
}

#[utoipa::path(
    tags = ["FF"],
    responses(
        (status = 200, description = "OK", body = bool),
    )
)]
#[post("/remove")]
pub async fn remove(
    data: web::Json<model::InConfigID>,
    core: web::Data<core::Core>,
) -> actix_web::Result<impl Responder> {
    core.ffdb
        .remove_ff_config(&data.server_uid, &data.config_uid)
        .await
        .map_err(|e| {
            actix_web::error::ErrorInternalServerError(format!("Fail to remove config: {}", e))
        })?;
    Ok(web::Json(true))
}

#[utoipa::path(
    tags = ["FF"],
    responses(
        (status = 200, description = "OK", body = model::ff::OutGetAllFFConfig),
    )
)]
#[get("/get_all_configs")]
pub async fn get_all_configs(core: web::Data<core::Core>) -> actix_web::Result<impl Responder> {
    let configs = core.ffdb.get_all_ff_configs().await.map_err(|e| {
        actix_web::error::ErrorInternalServerError(format!("Fail to get all configs: {}", e))
    })?;
    Ok(web::Json(model::ff::OutGetAllFFConfig { configs }))
}

#[utoipa::path(
    tags = ["FF"],
    responses(
        (status = 200, description = "OK", body = model::ff::OutGetFFStatus),
    )
)]
#[get("/get_all_status")]
pub async fn get_all_status(core: web::Data<core::Core>) -> actix_web::Result<impl Responder> {
    let status = core.ffdb.get_all_ff_status().await.map_err(|e| {
        actix_web::error::ErrorInternalServerError(format!("Fail to get all status: {}", e))
    })?;
    Ok(web::Json(model::ff::OutGetFFStatus { status }))
}

#[utoipa::path(
    tags = ["FF"],
    responses(
        (status = 200, description = "OK", body = model::ff::videoadaptationset::VideoRepresentation),
    )
)]
#[get("/create_default_video_representation")]
pub async fn create_default_video_representation() -> actix_web::Result<impl Responder> {
    Ok(web::Json(VideoRepresentation::new()))
}

#[utoipa::path(
    tags = ["FF"],
    responses(
        (status = 200, description = "OK", body = model::ff::videoadaptationset::VideoAdaptationSet),
    )
)]
#[get("/create_default_video_adaptation_set")]
pub async fn create_default_video_adaptation_set() -> actix_web::Result<impl Responder> {
    Ok(web::Json(VideoAdaptationSet::new()))
}

#[utoipa::path(
    tags = ["FF"],
    responses(
        (status = 200, description = "OK", body = String),
    )
)]
#[post("/get_ff_command")]
pub async fn get_ff_command(
    data: web::Json<model::InConfigID>,
    core: web::Data<core::Core>,
) -> actix_web::Result<impl Responder> {
    let ret = core
        .get_ff_cmd(&data.server_uid, &data.config_uid)
        .await
        .map_err(|e| {
            actix_web::error::ErrorInternalServerError(format!("Fail to get ff cmd: {}", e))
        })?;
    Ok(web::Json(ret))
}

#[utoipa::path(
    tags = ["FF"],
    responses(
        (status = 200, description = "OK", body = String),
    )
)]
#[post("/create")]
pub async fn create(
    data: web::Json<model::ff::InFFCreate>,
    core: web::Data<core::Core>,
) -> actix_web::Result<impl Responder> {
    let uid = core
        .ffdb
        .create_ff_config(&data.server_uid, &data.name, &data.output)
        .await
        .map_err(|e| {
            actix_web::error::ErrorInternalServerError(format!("Fail to create ff config: {}", e))
        })?;
    Ok(web::Json(uid))
}

pub fn service() -> impl HttpServiceFactory + OpenApiFactory + 'static {
    utoipa_actix_web::scope("/ff")
        .service(create_default_video_representation)
        .service(create_default_video_adaptation_set)
        .service(get_ff_command)
        .service(create)
        .service(update)
        .service(set_config_state)
        .service(remove)
        .service(get_all_configs)
        .service(get_all_status)
}
